File: src/distance/distance.js
var gpio = require('gpio');
var microtime = require('microtime');

var PIN = 12; // use this pin for the ultrasonic distance sensor

gpio12 = gpio.export(12, {
    direction: 'out',
    ready: function() {
        console.log(microtime.nowDouble() + ' - pin12 ready');
        if (gpio12.value == 1) {
            gpio12.reset(); // assure there is no signal on the pin.
        }
        setTimeout(getDistance, 1000);
        setTimeout(die, 10000);
    }
});


alpha = 0;
beta = 0;

gpio12.on('change', function(val) {
    console.log(microtime.nowDouble() + ' - change on pin12: ' + val);
    // if (gpio12.direction == 'inX') {
    if (val == 1) {
        alpha = microtime.nowDouble();
    } else if (val == 0) {
        beta = microtime.nowDouble();
        gamma = beta - alpha;
        cm = Math.round(34320.48 * gamma);
        console.log('CM: ' + cm + ', gamma: ' + gamma);
    }
    // }
    // if (val == 0) {
    //     setTimeout(getDistance, 1000);
    // }
});

getDistance = function() {
    alpha = 0;
    beta = 0;

    if (gpio12.direction == 'in') {
        gpio12.setDirection('out');
        while (microtime.nowDouble() < now + 0.00075) {}
    }

    var now = microtime.nowDouble();

    gpio12.set();
    while (microtime.nowDouble() < now + 0.00075) {}

    console.log(microtime.nowDouble() + ' - pin12 set');
    gpio12.setDirection('in');

    // console.log('val before while: ' + gpio12.value);
    //while (gpio12.value == 0) {
    // wait for 0, then start
    //}

    // alpha = microtime.nowDouble(); // start timing
    // delta = 0;
    // while (gpio12.value == 1) {

    // }
    // beta = microtime.nowDouble();
    // gamma = beta - alpha;
    // cm = Math.round(34320.48 * gamma);
    // console.log('CM: ' + cm + ', gamma: ' + gamma);
    //setTimeout(getDistance, 1000);
}


die = function() {
    gpio12.removeAllListeners('change');
    gpio12.reset();
    gpio12.unexport(function() {
        process.exit();
    });
}

File: src/distance/distance.py
#!/usr/bin/python
# speed of sound 1126 feet per second
#13512 inches per second 
# import time module
import time

# import gpio module
import RPi.GPIO as GPIO
GPIO.setmode(GPIO.BOARD)

count = 0
#loop used to check values multiple times
while (count < 100):

   alpha = 0
   beta = 0
   gamma = 0
   delta = 0
   inch = 0
   cm = 0
   
   # set up gpio pin 12 as output
   GPIO.setup(12, GPIO.OUT)

   # Be sure there is no signal to parallax ping: pin 11, 0 = False
   GPIO.output(12, 0)

   # send parallax ping a signal: pin 11, 1 = True
   GPIO.output(12, 1)
         
   time.sleep(0.00075)
   # set up gpio pin 12 as input
   GPIO.setup(12, GPIO.IN)

   
   #GPIO.input(12)

   # Start time
   alpha = time.time()

   # while GPIP.input(12) == 1 and delta < 20:
   while GPIO.input(12) == 1:
      # while the pin is high, wait.
      delta = delta + 1
   else:
      # Stop time
      beta = time.time()
      gamma = beta - alpha
      # print parallax ping value
      # calc for speed of sound at inches per second should be "correct"
      inch = 13512 * gamma
      #calc for speed at centimeters per second should be "correct"
      cm = 34320.48 * gamma
      #displays time, inches, cm should be the distance the object is from the device
      #print "Time: ", gamma
      #print "Inch: ", inch
      print "CM: ", round(cm), ' gamma: ', gamma
      #increase count by 1
   count = count + 1
   #sleep thread for 2 seconds
   time.sleep(1)
